File: java/897_IncreasingOrderSearchTree.java
class Solution {
    TreeNode curr;
    public TreeNode increasingBST(TreeNode root) {
        curr = new TreeNode(0);
        TreeNode ans = curr;
        inorder(root);
        return ans.right;
    }
    
    public void inorder(TreeNode root) {
        if (root == null) return;
        
        inorder(root.left);
        curr.right = root;
        curr = root;
        root.left = null;
        inorder(root.right);
    }
}

File: swift/897_IncreasingOrderSearchTree.swift
class Solution {
    func increasingBST(_ node: TreeNode?) -> TreeNode? {
        var root: TreeNode? = node
        var stack = [TreeNode?]()
        var head: TreeNode? = nil, curr: TreeNode? = nil
        while root != nil || !stack.isEmpty {
            while root != nil {
                stack.append(root)
                root = root?.left
            }
            root = stack.removeLast()
            if head == nil {
                curr = root
                head = root
            } else if curr != nil {
                curr?.right = root
                curr = root
            }
            curr?.left = nil
            root = root?.right
        }
        return head
    }
}
